sql: support scalar expressions (& placeholders) for alter range ... relocate

knz opened the pull request:
Fixes #73315.
See the individual commits for details.

knz committed:
sql: support scalar expressions (& placeholders) for relocate store IDs

This commit makes it possible to use arbitrary scalar
expressions (incl subqueries, placeholders etc) in the FROM/TO clauses
of ALTER RANGE ... RELOCATE statements.

For example:

```
ALTER RANGE 123 RELOCATE FROM $1 TO $2
ALTER RANGE 123 RELOCATE FROM $1 TO $1+10
```

Release note (sql change): The experimental ALTER RANGE...RELOCATE
syntax now accepts arbitrary scalar expressions as the source and target
store IDs.

Release note (sql change): The output of `EXPLAIN ALTER RANGE
... RELOCATE` now includes the source and target store IDs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73807)
<!-- Reviewable:end -->

knz commented:
@rafiss @otan do you have some yacc-fu that'd be useful here? By defining clauses as `FROM a_expr TO a_expr`,  we get a shift/reduce conflict on TO inside the interval syntax (e.g. INTERVAL DAYS TO MONTHS).  Is there a way to ask yacc to prefer shifting in that case?

otan commented:
you could try use `%prec` or maybe even consider b_expr or below (or force parenthesis around FROM/TO)

knz commented:
> you could try use %prec

can you explain how that works? Or maybe link to a doc that explains?

> or maybe even consider b_expr or below

unfortunately the interval syntax is in d_expr (the narrowest one)

> or force parenthesis around FROM/TO

yeah I thought about that, and if there's no solution via %prec, I'll put the choice to a vote on the KV team: either `TO y FROM x`, or `FROM (x) TO (y)`

otan commented:
Best I can find is https://www.ibm.com/docs/en/zos/2.3.0?topic=section-precedence-in-grammar-rules, I only have a vague understanding of it cargo culting from pg

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the syntax discussion

If we support placeholders now, it would be good to have some tests for that. I can't find many existing tests for RELOCATE, maybe we can add some similar to `TestSplitAt`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73807#-:-MquSb5I04z28wTJInP8:bdb1okh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @lunevalex)



<!-- Sent from Reviewable.io -->

lunevalex reviewed (approved):
LGTM

knz commented:
Thanks to the goyacc source code, I found out that it uses a super old algorithm that only supports _real_ token names in `%prec` (not pseudo-tokens) for the shifting rule. Once I understood that, I was able to fix the syntax.

knz commented:
For future reference, I troubleshooted my yacc problem with the following mini-grammar:

```yacc
%token RELOCATE FROM TO FOR SELECT DAY TO MINUTE

%left INTERVAL_SHORT
%left TO

%%

top:
   RELOCATE FROM a TO a FOR SELECT

a: b

b: DAY           %prec INTERVAL_SHORT
 | DAY TO MINUTE %prec TO
```

otan commented on pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported:
```diff
@@ -73,10 +73,10 @@ statement error operation is unsupported in multi-tenancy mode
 
 statement error operation is unsupported in multi-tenancy mode
-ALTER RANGE 1 RELOCATE FROM 1 TO 2
+ALTER RANGE 1 RELOCATE TO 2 FROM 1
```
revert?

knz replied:
Good call. Done.

knz committed:
sql: support scalar expressions for the simple range ID in relocate

This commit makes it possible to use arbitrary scalar
expressions (incl subqueries, placeholders etc) in the RANGE clause
of ALTER RANGE ... RELOCATE statements.

For example:

```
ALTER RANGE $1 RELOCATE FROM 3 TO 4
```

Release note (sql change): The experimental ALTER RANGE...RELOCATE
syntax now accepts arbitrary scalar expressions as the range ID
when the FOR clause is not used.

knz commented:
I have extended the PR with a 2nd commit that also adds placeholder support in the first position, so that this PR closes the linked issue entirely.

Please take another look.

otan reviewed (approved):
lgtm, some small test failure that look like simple rewrites

knz commented:
TFYRs!

bors r=lunevalex,otan

knz committed:
sql/opt/exec: include replica subject in EXPLAIN ALTER RANGE RELOCATE

Release note (sql change): The output of `EXPLAIN ALTER RANGE
... RELOCATE` now includes which replicas are subject to the relocation.

knz commented:
bors r=lunevalex,otan

knz commented:
bors r=lunevalex,otan

RaduBerinde commented:
Opt changes LGTM.





<!-- Sent from Reviewable.io -->
